In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
data = df_market_data[['price_change_percentage_24h','price_change_percentage_7d','price_change_percentage_14d','price_change_percentage_30d','price_change_percentage_60d','price_change_percentage_200d','price_change_percentage_1y']]
scaled_data = StandardScaler().fit_transform(data)

In [6]:
# Create a DataFrame with the scaled data
scaled_df = pd.DataFrame(scaled_data,columns=['price_change_percentage_24h','price_change_percentage_7d','price_change_percentage_14d','price_change_percentage_30d','price_change_percentage_60d','price_change_percentage_200d','price_change_percentage_1y'])

# Copy the crypto names from the original data
# Set the coinid column as index
scaled_df['coinid'] = df_market_data.index
scaled_df = scaled_df.set_index('coinid')
# Display sample data
scaled_df.head()

price_change_percentage_24h  price_change_percentage_7d  \
coinid                                                                  
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coinid                                                                   
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coinid                                                                    
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coinid                                    
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [7]:
# Create a list with the number of k-values from 1 to 11
k_values_list = range(1,11)

In [8]:
# Create an empty list to store the inertia values
inertia_values = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for k in k_values_list:
    k_model =KMeans(n_clusters = k , random_state = 1)
    k_model.fit(scaled_df)
    inertia_values.append(k_model.inertia_)

In [9]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_dict = {
    "k values": k_values_list,
    "inertal values": inertia_values
}

# Create a DataFrame with the data to plot the Elbow curve
elbow_data_df = pd.DataFrame(elbow_data_dict)

In [10]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_curve = elbow_data_df.hvplot.line(
    x="k values",
    y= "inertal values",
    title= "Elbow Curve"
)
elbow_curve

:Curve   [k values]   (inertal values)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:**  From the plot, the elbow point appears to be at k = 4. This is where the curve shows a noticeable bend or elbow, after which the decrease in inertia becomes more gradual. Therefore, k = 4 is the best value because it balances between minimizing inertia and avoiding unnecessary complexity by adding more clusters.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [32]:
# Initialize the K-Means model using the best value for k
k_means_model = KMeans(n_clusters=4, random_state=1)

In [33]:
# Fit the K-Means model using the scaled data
k_means_model.fit(scaled_df)

KMeans(n_clusters=4, random_state=1)

In [34]:
# Predict the clusters to group the cryptocurrencies using the scaled data
clusters = k_means_model.predict(scaled_df)

# Print the resulting array of cluster values.
print(clusters)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [14]:
# Create a copy of the DataFrame
scaled_df_copy = scaled_df.copy()

In [15]:
# Add a new column to the DataFrame with the predicted clusters
scaled_df_copy['Cluster segments'] = clusters

# Display sample data
scaled_df_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coinid                                                                  
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coinid                                                                   
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coinid                                                                    
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Cluster segments  
coinid                                                      
bitcoin                        -0.251637                 2  
ethereum                       -0.199352                 2  
tether                         -0.282061                 0  
ripple                         -0.295546                 0  
bitcoin-cash                   -0.270317                 2

In [16]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

cluster_plot = scaled_df_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    c="Cluster segments",  
    hover_cols=["coinid"],
    title= 'Cluster Scatter Plot'  
)
cluster_plot


:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster segments,coinid)

---

### Optimize Clusters with Principal Component Analysis.

In [17]:
# Create a PCA model instance and set `n_components=3`.
pca_model = PCA(n_components=3)

In [18]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

pca_data = pca_model.fit_transform(scaled_df)

# View the first five rows of the DataFrame. 
pca_data[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [19]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca_model.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance is gained through the sum of the avbove three values .
0.3719856+0.34700813+0.17603793=0.89503166
So, the total explained variance of the three principal components is 0.895 (or 89.5%)

In [20]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
# Copy the crypto names from the original data

pca_df = pd.DataFrame(pca_data,columns=['PC1','PC2','PC3'])

# Set the coinid column as index
pca_df["coinid"] = scaled_df.index
pca_df = pca_df.set_index('coinid')

# Display sample data
pca_df.head()


PC1       PC2       PC3
coinid                                    
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [21]:
# Create a list with the number of k-values from 1 to 11
k_list =range(1,11)

In [22]:
# Create an empty list to store the inertia values
inertia_value_list = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in k_list:
    kmodel_pca = KMeans(n_clusters=k,random_state=1)
    kmodel_pca.fit(pca_df)
    inertia_value_list.append(kmodel_pca.inertia_)

In [23]:
# Create a dictionary with the data to plot the Elbow curve
elbow_dict_pca = {
    'k values':k_list,
    'inertia values':inertia_value_list
}
# Create a DataFrame with the data to plot the Elbow curve
elbow_df_pca = pd.DataFrame(elbow_dict_pca)

In [24]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_curve_pca = elbow_df_pca.hvplot.line(
    x='k values',
    y='inertia values',
    title = 'Elbow Curve_PCA'
)
elbow_curve_pca

:Curve   [k values]   (inertia values)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** k = 4 is where the curve shows a noticeable bend or elbow in ths plot as well, after which the decrease in inertia becomes more gradual. Therefore, k = 4 is the best value because it balances between minimizing inertia and avoiding unnecessary complexity by adding more clusters.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** k value did not change

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [25]:
# Initialize the K-Means model using the best value for k
model_pca = KMeans(n_clusters=4,random_state=1)

In [26]:
# Fit the K-Means model using the PCA data
model_pca.fit(pca_df)

KMeans(n_clusters=4, random_state=1)

In [27]:
# Predict the clusters to group the cryptocurrencies using the PCA data
clusters_pca = model_pca.predict(pca_df)
# Print the resulting array of cluster values.
print(clusters_pca)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


In [28]:
# Create a copy of the DataFrame with the PCA data
pca_df_copy = pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_df_copy['Cluster segments'] = clusters_pca
# Display sample data
pca_df_copy.head()

PC1       PC2       PC3  Cluster segments
coinid                                                      
bitcoin      -0.600667  0.842760  0.461595                 2
ethereum     -0.458261  0.458466  0.952877                 2
tether       -0.433070 -0.168126 -0.641752                 0
ripple       -0.471835 -0.222660 -0.479053                 0
bitcoin-cash -1.157800  2.041209  1.859715                 2

In [29]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

pca_cluster_plot = pca_df_copy.hvplot.scatter(
    x='PC1',
    y='PC2',
    c='Cluster segments',
    hover_cols=['coinid'],
    title='Cluster Scatter Plot-PCA'
)
pca_cluster_plot


:Scatter   [PC1]   (PC2,Cluster segments,coinid)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves
composite_elbow_curve =elbow_curve + elbow_curve_pca
composite_elbow_curve

:Layout
   .Curve.I  :Curve   [k values]   (inertal values)
   .Curve.II :Curve   [k values]   (inertia values)

In [31]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

composite_cluster_plot = cluster_plot + pca_cluster_plot
composite_cluster_plot

:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Cluster segments,coinid)
   .Scatter.II :Scatter   [PC1]   (PC2,Cluster segments,coinid)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 
 
When fewer features are used to cluster the data (such as through PCA), the following impacts can be observed:

1. **Elbow Curve Comparison:**
   - In both the regular and PCA elbow curves, the inertia decreases as the number of clusters increases. However, the PCA elbow curve shows a sharper drop in inertia with fewer clusters. This suggests that using fewer features (through PCA) allows for a more efficient clustering process, as the model needs fewer clusters to achieve a similar level of inertia reduction.
   - The elbow point in both graphs occurs around 3-4 clusters, indicating that additional clusters beyond this point do not significantly reduce inertia.

2. **Cluster Scatter Plot Comparison:**
   - The scatter plots show a clearer separation of data points in the PCA-based clustering. The data points are more compactly grouped in the two principal components (PC1, PC2), compared to the original feature space where the points are more dispersed.
   - Using fewer features (PCA) simplifies the dimensionality, making the clusters easier to visualize and analyze. However, it may cause some loss of information since the original feature space captures more variability.

In summary, using fewer features (via PCA) can streamline the clustering process by reducing complexity, but it may also cause some information loss, potentially affecting the granularity of the clusters. The clustering becomes more efficient and distinct, as seen in the scatter plots and elbow curves.